### PR TITLE
[AMD]feat: Implement fused qk_norm and RoPE caching for improving Qwen3.5 BF16

### PR DESCRIPTION
**Pull request by zhentaocc**
## Motivation

Integrate AIter's fused QK norm + RoPE + KV cache kernel into Qwen3.5 attention on ROCm/HIP to improve inference performance. Qwen3.5 uses a hybrid attention architecture with **partial rotary embeddings** (`partial_rotary_factor=0.25`, i.e., `rotary_dim=64` for `head_dim=256`). This PR wires SGLang to the AIter kernels that support `rotary_dim`, enabling correct partial RoPE handling while fusing RMSNorm, RoPE application, and KV cache writes into a single kernel launch.

**Upstream dependency**: [ROCm/aiter PR #2199](https://github.com/ROCm/aiter/pull/2199) adds the `rotary_dim` parameter to `fused_qk_norm_rope_cache_pts_quant_shuffle` and `fused_qk_norm_mrope_3d_cache_pts_quant_shuffle`. This SGLang change requires that PR (or equivalent) to be merged and available in the AIter package.

## Usage/How to enable

upgrade aiter to latest when  [ROCm/aiter PR #2199](https://github.com/ROCm/aiter/pull/2199) is merged and run:
```bash
export SGLANG_FUSED_QK_NORM_ROPE_CACHE_PTS_QUANT_SHUFFLE=1
sglang serve     --attention-backend triton     --model-path $MODEL     --host=0.0.0.0     --port $PORT     --tensor-parallel-size $TP     --trust-remote-code     --mem-fraction-static 0.8 --enable-fused-qk-norm-rope   
```

### `python/sglang/srt/models/qwen3_5.py`

- **`_fused_qk_norm_rope_cache_quant_aiter()`**: New method that invokes AIter fused kernels:
  - For standard RoPE: `aiter.fused_qk_norm_rope_cache_pts_quant_shuffle()` with `rotary_emb.rotary_dim`
  - For MRoPE (multi-dimensional RoPE): `aiter.fused_qk_norm_mrope_3d_cache_pts_quant_shuffle()` with `rotary_emb.rotary_dim`
- **`self_attention()`**: When `enable_fused_qk_norm_rope_set_kv_aiter()` returns true and `is_gfx95_supported()`, uses the fused path instead of separate `_apply_qk_norm`, `rotary_emb`, and attention KV cache writes. Passes `save_kv_cache=False` to the attention backend since the fused kernel writes directly to the KV cache.

- **Reshuffle QKV gate weights (full attention only)** (Removed from this PR for easier review):
  - **`_reshuffle_qkv_gate`**: Flag enabled when `SGLANG_USE_AITER=1` on HIP; gates the reshuffle path.
  - **`_reshuffle_qkv_proj_weight_to_q_k_v_gate()`**: Converts `qkv_proj` weight from checkpoint layout `[q_gate, k, v]` (interleaved `[q0, g0, q1, g1, ..., k, v]`) to `[q, k, v, gate]` (`[q0, q1, ..., k, v, g0, g1, ...]`) for AIter kernel compatibility.
  - **`self_attention()`**: When `_reshuffle_qkv_gate` and `attn_output_gate`, splits `qkv` as `[q, k, v, gate]` instead of the interleaved `[q_gate, k, v]` layout.

### `python/sglang/srt/models/utils.py`

- **`enable_fused_set_kv_buffer_hip()`**: New helper to enable fused set_kv only on HIP with bf16/fp8 KV cache (excludes SWAKVPool).
- **`enable_fused_qk_norm_rope_set_kv_aiter()`**: Gating logic for the fused path. Requires:
  - `enable_fused_qk_norm_rope` (server arg)
  - `SGLANG_USE_AITER=1` (env)
  - `SGLANG_FUSED_QK_NORM_ROPE_CACHE_PTS_QUANT_SHUFFLE=1` (env)
  - `enable_fused_set_kv_buffer_hip(forward_batch)` (HIP + bf16/fp8 KV cache)

## Accuracy Tests

- **Upstream AIter tests**: [ROCm/aiter PR #2199](https://github.com/ROCm/aiter/pull/2199)  validating partial rotary correctness against a pure-PyTorch reference for Qwen3.5-style configs (`head_size=256`, `rotary_dim=64`, etc.). 
- **SGLang-side**: 

After
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9674|±  |0.0049|
|     |       |strict-match    |     5|exact_match|↑  |0.9682|±  |0.0048|

## Benchmarking 
- **AIter kernel benchmarks**  [ROCm/aiter PR #2199](https://github.com/ROCm/aiter/pull/2199): Fused kernel shows ~15–30% latency reduction vs main on MI355X for typical Qwen3.5 shapes (e.g., T=32–1024, Hq=32, Hk=4, D=256). Please refer to this PR for detailed kernel perf comparison.
- **SGLang end-to-end**: Compare decoding throughput/latency on ROCm with `enable_fused_qk_norm_rope=True` vs `False` (with `SGLANG_USE_AITER=1`). TBD for Qwen3.5 before and after.
-
Before
```bash
 sglang serve     --attention-backend triton     --model-path $MODEL     --host=0.0.0.0     --port $PORT     --tensor-parallel-size $TP     --trust-remote-code     --mem-fraction-static 0.8 
```
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     160       
Benchmark duration (s):                  192.69    
Total input tokens:                      147038    
Total input text tokens:                 147038    
Total generated tokens:                  147952    
Total generated tokens (retokenized):    147623    
Request throughput (req/s):              0.83      
Input token throughput (tok/s):          763.08    
Output token throughput (tok/s):         767.82    
Peak output token throughput (tok/s):    864.00    
Peak concurrent requests:                21        
Total token throughput (tok/s):          1530.90   
Concurrency:                             15.39     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18535.68  
Median E2E Latency (ms):                 18662.69  
P90 E2E Latency (ms):                    20331.16  
P99 E2E Latency (ms):                    20857.56  
---------------Time to First Token----------------
Mean TTFT (ms):                          128.35    
Median TTFT (ms):                        97.52     
P99 TTFT (ms):                           409.59    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          19.92     
Median TPOT (ms):                        20.06     
P99 TPOT (ms):                           20.38     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           19.93     
Median ITL (ms):                         18.77     
P95 ITL (ms):                            19.03     
P99 ITL (ms):                            99.89     
Max ITL (ms):                            405.48    
==================================================
```
After:
```bash
export SGLANG_FUSED_QK_NORM_ROPE_CACHE_PTS_QUANT_SHUFFLE=1
sglang serve     --attention-backend triton     --model-path $MODEL     --host=0.0.0.0     --port $PORT     --tensor-parallel-size $TP     --trust-remote-code     --mem-fraction-static 0.8 --enable-fused-qk-norm-rope   
```
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     160       
Benchmark duration (s):                  185.48    
Total input tokens:                      147038    
Total input text tokens:                 147038    
Total generated tokens:                  147952    
Total generated tokens (retokenized):    147301    
Request throughput (req/s):              0.86      
Input token throughput (tok/s):          792.74    
Output token throughput (tok/s):         797.66    
Peak output token throughput (tok/s):    896.00    
Peak concurrent requests:                21        
Total token throughput (tok/s):          1590.40   
Concurrency:                             15.39     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   17846.62  
Median E2E Latency (ms):                 17972.05  
P90 E2E Latency (ms):                    19588.60  
P99 E2E Latency (ms):                    20075.13  
---------------Time to First Token----------------
Mean TTFT (ms):                          125.43    
Median TTFT (ms):                        93.01     
P99 TTFT (ms):                           419.89    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          19.18     
Median TPOT (ms):                        19.31     
P99 TPOT (ms):                           19.59     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           19.19     
Median ITL (ms):                         18.07     
P95 ITL (ms):                            18.38     
P99 ITL (ms):                            95.11     
Max ITL (ms):                            406.37    
==================================================
```

## Profiling
Before
image

After

image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
5. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/utils.py [hunk 1]
```diff
@@ -116,6 +121,31 @@ def enable_fused_set_kv_buffer(forward_batch: ForwardBatch):
...
+    """Enable aiter fused qk_norm + RoPE/MRoPE + set_kv on HIP with bf16 KV cache."""
+    _use_aiter = is_hip() and get_bool_env_var("SGLANG_USE_AITER")
+    from sglang.srt.server_args import get_global_server_args
+
+    return (
+        get_global_server_args().enable_fused_qk_norm_rope
+        and _use_aiter
+        and enable_fused_set_kv_buffer_hip(forward_batch)
+        and get_bool_env_var("SGLANG_FUSED_QK_NORM_ROPE_CACHE_PTS_QUANT_SHUFFLE")
+        # TODO: default to False if not set, and remove the default value from the env when aiter changes are included
```
Comment (zhentaocc) [inline python/sglang/srt/models/utils.py:311]: @yichiche added an argument to disable the path by default now. No need to wait for new aiter version ... Otherwise, it's impossible to use these changes in next official release.

Comment (zhentaocc) [reply python/sglang/srt/models/utils.py:311]: I also add explicit error message to handle incompatible use, this won't break anything. Just update the default value when new aiter is ready in next SGLang.

File: python/sglang/srt/models/qwen3_5.py [hunk 2]
```diff
@@ -806,9 +960,22 @@ def self_attention(
         else:
             q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
 
-        q, k = self._apply_qk_norm(q, k)
-        q, k = self.rotary_emb(positions, q, k)
-        attn_output = self.attn(q, k, v, forward_batch)
+        save_kv_cache = True
```
Comment (yichiche) [inline python/sglang/srt/models/qwen3_5.py:978]: The fused path condition needs an additional guard:

```python
if (
    enable_fused_qk_norm_rope_set_kv_aiter(forward_batch)
    and is_gfx95_supported()
    and not self.attn_output_gate   # <-- add this
):
```

**Why:** When `attn_output_gate=True` (the default for all Qwen3.5 full-attention layers), `qkv_proj` outputs `total_num_heads * 2` Q-heads worth of data — Q and Gate interleaved per head — followed by K and V:

```
qkv layout:  [Q0, G0, Q1, G1, ..., Q31, G31,  K0..K3,  V0..V3]
              |---------- 64×256 ----------|  |4×256|  |4×256|
                    = 16384                     1024     1024
```

But `_fused_qk_norm_rope_cache_quant_aiter` passes `num_heads_q=self.num_heads` (=32, not 64) to the AIter kernel. The kernel addresses the QKV tensor as `qkv[token_id * (num_heads_q + num_heads_k + num_heads_v) + head_id] * HEAD_SIZE` (see [`rope_common.h` L4368](https://github.com/ROCm/aiter/blob/main/csrc/kernels/rope/rope_common.h)), so it expects a contiguous `[Q(32×256), K(4×256), V(4×256)]` layout totaling 10240 elements — but the actual tensor has 18432 elements. The kernel reads corrupted K/V from the middle of the Q+Gate region and writes them directly to KV cache (`save_kv_cache=False`).

Since the QKV gate reshuffling logic (`_reshuffle_qkv_proj_weight_to_q_k_v_gate`) was intentionally removed from this PR, the fused path **must not** run when `attn_output_gate` is active. Without this guard, every standard Qwen3.5 full-attention layer produces incorrect attention output.

**Review by yichiche — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/qwen3_5.py (hunk 2 above)
Comment (zhentaocc) [reply python/sglang/srt/models/qwen3_5.py:978]: Nice catch! Instead, I propose to add 
`qkv = torch.cat([q, k, v], dim=-1)`
when `self.attn_output_gate is not None` .

Comment (zhentaocc) [reply python/sglang/srt/models/qwen3_5.py:978]: To not influence existing path introducing extra concat copy kernel, add the concat op when `fusion=True and attention_gate_ouput is not None`

Comment (yichiche) [reply python/sglang/srt/models/qwen3_5.py:978]: LGTM, feel free to proceed

**Review by yichiche — APPROVED** [review]
LGTM. Since we are using a flag and a global variable for control, it is fine to merge either before or after the aiter upgrade.

**Comment by 1am9trash** [comment]
@amd-bot ci-status

**Comment by hubertlu-tw** [comment]
Hi @zhentaocc,
I tested the PR locally on a MI355 server.
 
No improvement at 8192/8192 and 1024/8192. The fused path is again slightly slower across the board. The regression is slightly worse for TTFT (6-12%) because the fused path now affects the prefill stage directly, but since 1024 tokens of prefill is still small, the overhead (extra qkv.contiguous() copy, per-call 1.0 + weight computation, Python-level gating) exceeds the kernel launch savings.



```
SGLANG_FUSED_QK_NORM_ROPE_CACHE_PTS_QUANT_SHUFFLE=1 SGLANG_USE_AITER_UNIFIED_ATTN=1 SGLANG_USE_AITER=1 python3 -m sglang.launch_server --model-path /data2/Qwen/Qwen3.5-397B-A17B-FP8/ --tp 8 --attention-backend aiter --trust-remote-code --model-loader-extra-config '{"enable_multithread_load": true}' --watchdog-timeout 1200 --mem-fraction-static 0.9 --host 0.0.0.0 --port 9000 --disable-radix-cache --enable-aiter-allreduce-fusion --max-running-requests 128 --page-size 16 --enable-fused-qk-norm-rope
```


```
python3 benchmark_serving.py \
    --model "/data2/Qwen/Qwen3.5-397B-A17B-FP8" \
    --base-url "[http://0.0.0.0:9000"](http://0.0.0.0:9000%22) \
    --dataset-name random \
    --random-input-len 1024 \
    --random-output-len 8192 \
    --random-range-ratio 1.0 \
    --num-prompts 64 \
    --max-concurrency 16 \
    --request-rate inf \
    --ignore-eos \
    --save-result \
    --num-warmups 32 \
    --percentile-metrics 'ttft,tpot,itl,e2el' \
    --result-dir "." \
    --result-filename "case_b_1024in_8192out.json"
```


Results: 1024 in / 8192 out (concurrency=16)
image

**Comment by hnyls2002** [comment]
Thanks @zhentaocc! The last benchmark on this PR reports no improvement at 1024/8192 and 8192/8192 with a 6-12% TTFT regression, attributed to the extra `qkv.contiguous()` copy and per-call `1.0 + weight`, so the stated perf motivation is not currently met; the PR also still carries `amd-aiter-not-ready`. Closing as parked - please reopen when the aiter kernel path shows a win.
